### PR TITLE
feat: Add an experimental lambda-based materialization engine

### PR DESCRIPTION
**Pull request by achals**
Signed-off-by: Achal Shah <achals@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: sdk/python/feast/repo_config.py [hunk 1]
```diff
@@ -205,6 +213,22 @@ def online_store(self):
...
+            if isinstance(self._batch_engine_config, Dict):
+                self._batch_engine = get_batch_engine_config_from_type(
+                    self._batch_engine_config["type"]
+                )(**self._batch_engine_config)
+            elif isinstance(self._online_config, str):
+                self._batch_engine = get_batch_engine_config_from_type(
+                    self._online_config
+                )()
+            elif self._online_config:
+                self._batch_engine = self._batch_engine
```
Comment (kevjumba) [inline sdk/python/feast/repo_config.py:228]: what is this doing

File: sdk/python/feast/infra/materialization/lambda/app.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            )
+
+        join_key_to_value_type = {
+            entity.name: entity.dtype.to_value_type()
+            for entity in feature_view.entity_columns
+        }
+
+        written_rows = 0
+
+        for batch in table.to_batches(DEFAULT_BATCH_SIZE):
```
[the remark this answers is not in the document]
Comment (achals) [reply sdk/python/feast/infra/materialization/lambda/app.py:68]: It's that setting in the local engine - no specific reason.

File: sdk/python/feast/infra/materialization/lambda/app.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            f.write(config_bytes)
+
+        # Initialize the feature store
+        store = FeatureStore(repo_path=str(repo_path.resolve()))
+
+        view_name = event["view_name"]
+        view_type = event["view_type"]
+        path = event["path"]
+
+        bucket = path[len("s3://") :].split("/", 1)[0]
```
[the remark this answers is not in the document]
Comment (achals) [reply sdk/python/feast/infra/materialization/lambda/app.py:44]: Yeah sure

File: sdk/python/feast/infra/materialization/lambda/app.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        key = path[len("s3://") :].split("/", 1)[1]
+        print(f"Inferred Bucket: `{bucket}` Key: `{key}`", flush=True)
+
+        if view_type == "batch":
+            feature_view = store.get_feature_view(view_name)
```
[the remark this answers is not in the document]
Comment (achals) [reply sdk/python/feast/infra/materialization/lambda/app.py:49]: sure

File: sdk/python/feast/repo_config.py (hunk 1 above)
Comment (achals) [reply sdk/python/feast/repo_config.py:228]: Oops there's some bugs here

File: sdk/python/feast/infra/materialization/lambda/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (achals) [reply sdk/python/feast/infra/materialization/lambda/app.py:44]: Actually if this fails the lambda will quit with the stack trace so I'm okay leaving it as is

File: sdk/python/feast/infra/materialization/lambda/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (achals) [reply sdk/python/feast/infra/materialization/lambda/app.py:49]: I just added a TODO for now

File: sdk/python/feast/infra/feature_servers/aws_lambda/Dockerfile [hunk 5]
```diff
@@ -20,7 +20,6 @@ COPY README.md README.md
 # https://github.com/pypa/setuptools_scm#usage-from-docker
 # I think it also assumes that this dockerfile is being built from the root of the directory.
 RUN --mount=source=.git,target=.git,type=bind pip3 install --no-cache-dir -e '.[aws,redis]'
-RUN pip3 install -r sdk/python/feast/infra/feature_servers/aws_lambda/requirements.txt --target "${LAMBDA_TASK_ROOT}"
```
Comment (kevjumba) [inline sdk/python/feast/infra/feature_servers/aws_lambda/Dockerfile:23]: why is this being removed?

**Review by kevjumba — APPROVED** [review]

**Pull request merged**